File: sim/falsifier.go
package sim

import "fmt"

// Falsifier is a price-path predicate evaluated against one synthetic path.
//
// A path is a level series normalised to path[0] == 1, length horizon+1. Only
// the predicate sees it; no path leaves this package.
type Falsifier interface {
	Trips(path []float64) bool
	Kind() string
	Spec() map[string]any
	Validate() error
}

// TouchBelow trips if the level falls Drop below the start at any point.
// Drop is expressed as a positive fraction: 0.12 means "down 12%".
type TouchBelow struct{ Drop float64 }

func (f TouchBelow) Kind() string         { return "touch_below" }
func (f TouchBelow) Spec() map[string]any { return map[string]any{"kind": f.Kind(), "drop": f.Drop} }
func (f TouchBelow) Validate() error      { return checkFrac("drop", f.Drop) }
func (f TouchBelow) Trips(path []float64) bool {
	threshold := 1 - f.Drop
	for _, lvl := range path {
		if lvl <= threshold {
			return true
		}
	}
	return false
}

// TouchAbove trips if the level rises Rise above the start at any point.
type TouchAbove struct{ Rise float64 }

func (f TouchAbove) Kind() string         { return "touch_above" }
func (f TouchAbove) Spec() map[string]any { return map[string]any{"kind": f.Kind(), "rise": f.Rise} }
func (f TouchAbove) Validate() error {
	if f.Rise < 0 {
		return errInvalid("rise must be >= 0")
	}
	return nil
}
func (f TouchAbove) Trips(path []float64) bool {
	threshold := 1 + f.Rise
	for _, lvl := range path {
		if lvl >= threshold {
			return true
		}
	}
	return false
}

// TerminalBelow trips on the horizon date only. "Closes below" rather than
// "trades below" — a materially different and usually much harder falsifier,
// which is exactly the kind of difference null_probability is there to expose.
type TerminalBelow struct{ Drop float64 }

func (f TerminalBelow) Kind() string         { return "terminal_below" }
func (f TerminalBelow) Spec() map[string]any { return map[string]any{"kind": f.Kind(), "drop": f.Drop} }
func (f TerminalBelow) Validate() error      { return checkFrac("drop", f.Drop) }
func (f TerminalBelow) Trips(path []float64) bool {
	return path[len(path)-1] <= 1-f.Drop
}

// TerminalAbove trips on the horizon date only.
type TerminalAbove struct{ Rise float64 }

func (f TerminalAbove) Kind() string         { return "terminal_above" }
func (f TerminalAbove) Spec() map[string]any { return map[string]any{"kind": f.Kind(), "rise": f.Rise} }
func (f TerminalAbove) Validate() error {
	if f.Rise < 0 {
		return errInvalid("rise must be >= 0")
	}
	return nil
}
func (f TerminalAbove) Trips(path []float64) bool {
	return path[len(path)-1] >= 1+f.Rise
}

// DrawdownExceeds trips if peak-to-trough drawdown within the horizon exceeds
// Limit.
type DrawdownExceeds struct{ Limit float64 }

func (f DrawdownExceeds) Kind() string { return "drawdown_exceeds" }
func (f DrawdownExceeds) Spec() map[string]any {
	return map[string]any{"kind": f.Kind(), "limit": f.Limit}
}
func (f DrawdownExceeds) Validate() error { return checkFrac("limit", f.Limit) }
func (f DrawdownExceeds) Trips(path []float64) bool {
	return maxDrawdown(path) >= f.Limit
}

func checkFrac(name string, v float64) error {
	if v <= 0 || v >= 1 {
		return errInvalid(fmt.Sprintf("%s must be in (0,1), got %g", name, v))
	}
	return nil
}

// maxDrawdown is the largest peak-to-trough decline on the path, as a fraction.
func maxDrawdown(path []float64) float64 {
	peak := path[0]
	worst := 0.0
	for _, lvl := range path {
		if lvl > peak {
			peak = lvl
		}
		if dd := (peak - lvl) / peak; dd > worst {
			worst = dd
		}
	}
	return worst
}

File: co_agent/sim/falsifier.py
"""Price-path falsifier predicates.

A predicate is evaluated against a batch of synthetic paths at once: ``levels``
is a ``(paths, horizon + 1)`` array of price levels normalised so that
``levels[:, 0] == 1``.  Only the predicate sees it; no path leaves this package.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol, runtime_checkable

import numpy as np

from .params import SimInputError


@runtime_checkable
class Falsifier(Protocol):
    """A price-path predicate."""

    def trips(self, levels: np.ndarray) -> np.ndarray:
        """Return a boolean array, one entry per path."""

    @property
    def kind(self) -> str: ...

    def spec(self) -> dict[str, object]:
        """Provenance for ``sim_params.falsifier``."""

    def validate(self) -> None:
        """Raise :class:`SimInputError` if the predicate is malformed."""


def _check_fraction(name: str, value: float) -> None:
    if not 0 < value < 1:
        raise SimInputError(f"{name} must be in (0,1), got {value!r}")


def _check_non_negative(name: str, value: float) -> None:
    if value < 0:
        raise SimInputError(f"{name} must be >= 0, got {value!r}")


@dataclass(frozen=True, slots=True)
class TouchBelow:
    """Trips if a daily close falls ``drop`` below the start, on any day.

    ``drop`` is a positive fraction: 0.12 means "down 12%".

    **This is not an intraday touch.** Synthetic paths are close-to-close, so
    "any point" means any daily close, not any trade. A live thesis worded
    "trades 12% below" and resolved against intraday lows will trip more often
    than this null predicts, because a daily range straddles its close --
    ``null_probability`` would be biased low, the gate would misjudge it, and
    FR7's calibration net of it would be wrong in the flattering direction.

    Two ways to keep the null and the resolver in agreement: word falsifiers
    against closes, or feed the simulator OHLC bars and model the daily low.
    Whichever is chosen, the resolver must use the same definition -- the
    mismatch is silent.
    """

    drop: float

    @property
    def kind(self) -> str:
        return "touch_below"

    def spec(self) -> dict[str, object]:
        return {"kind": self.kind, "drop": self.drop}

    def validate(self) -> None:
        _check_fraction("drop", self.drop)

    def trips(self, levels: np.ndarray) -> np.ndarray:
        return (levels <= 1 - self.drop).any(axis=1)


@dataclass(frozen=True, slots=True)
class TouchAbove:
    """Trips if a daily close rises ``rise`` above the start, on any day.

    Close-based, not an intraday touch -- see :class:`TouchBelow`.
    """

    rise: float

    @property
    def kind(self) -> str:
        return "touch_above"

    def spec(self) -> dict[str, object]:
        return {"kind": self.kind, "rise": self.rise}

    def validate(self) -> None:
        _check_non_negative("rise", self.rise)

    def trips(self, levels: np.ndarray) -> np.ndarray:
        return (levels >= 1 + self.rise).any(axis=1)


@dataclass(frozen=True, slots=True)
class TerminalBelow:
    """Trips on the closing price of the horizon date only.

    Closing *on* the date rather than closing below it at any point during the
    horizon -- a materially different and usually much harder falsifier, which is
    exactly the kind of difference ``null_probability`` exists to expose.
    """

    drop: float

    @property
    def kind(self) -> str:
        return "terminal_below"

    def spec(self) -> dict[str, object]:
        return {"kind": self.kind, "drop": self.drop}

    def validate(self) -> None:
        _check_fraction("drop", self.drop)

    def trips(self, levels: np.ndarray) -> np.ndarray:
        return levels[:, -1] <= 1 - self.drop


@dataclass(frozen=True, slots=True)
class TerminalAbove:
    """Trips on the horizon date only."""

    rise: float

    @property
    def kind(self) -> str:
        return "terminal_above"

    def spec(self) -> dict[str, object]:
        return {"kind": self.kind, "rise": self.rise}

    def validate(self) -> None:
        _check_non_negative("rise", self.rise)

    def trips(self, levels: np.ndarray) -> np.ndarray:
        return levels[:, -1] >= 1 + self.rise


@dataclass(frozen=True, slots=True)
class DrawdownExceeds:
    """Trips if peak-to-trough drawdown within the horizon exceeds ``limit``."""

    limit: float

    @property
    def kind(self) -> str:
        return "drawdown_exceeds"

    def spec(self) -> dict[str, object]:
        return {"kind": self.kind, "limit": self.limit}

    def validate(self) -> None:
        _check_fraction("limit", self.limit)

    def trips(self, levels: np.ndarray) -> np.ndarray:
        return max_drawdown(levels) >= self.limit


def max_drawdown(levels: np.ndarray) -> np.ndarray:
    """Largest peak-to-trough decline per path, as a fraction."""
    peak = np.maximum.accumulate(levels, axis=1)
    return ((peak - levels) / peak).max(axis=1)
